plugin/pulpcore/plugin/changeset/model.py

```diff
@@ -405,3 +407,66 @@ def save(self):
 
     def __hash__(self):
         return hash(self.relative_path)
+
+
+class NopDownloader(BaseDownloader):
+    """
+    A no-operation (NOP) downloader.
+    """
+
+    def __init__(self):
+        super().__init__('')
+
+    def _send(self):
+        pass
+
+    async def run(self):
+        pass
+
+
+class NopPendingArtifact(PendingArtifact):
+    """
+    No operation (NOP) pending artifact.
+    """
+
+    def __init__(self, content):
+        """
+        Args:
+            content (PendingContent): The associated pending content.
+        """
+        super().__init__(Artifact(), '', '')
+        self.content = content
+
+    def downloader(self):
+        return NopDownloader()
+
+    def settle(self):
+        pass
+
+    def save(self):
+        pass
+
+
+class DownloadTask(asyncio.Task):
```

Comment:
I was just reading about inheriting from `asyncio.Task` and this does use Futures to run co-routines. What I'm gathering from this is that the changeset uses Futures on it's own, outside of the futures based downloaders. Or is the main benefit that this allows for the callback opportunity?

Reply:
Also the docs say `Don’t directly create Task instances: use the ensure_future() function or the AbstractEventLoop.create_task() method`. It sounds like we shouldn't be subclassing this. Am I reading it right?

Reply:
The _asyncio_ API is a bit schizophrenic with regards to `Future` and `Task`.  It cannot decide if it's procedural or object-oriented.  The `Task` class is exposed and the API is fairly `Future` oriented and exposing and documenting a class then saying don't use it does not make sense.  That said, the warning against instantiating a `Task` directly is clear (though I think unfounded) and does preclude us from extending it (which is unfortunate).

I pushed another approach which uses `ensure_future()` to create the task, then adds the callback.  The `DownloadTask` included a `artifact` attribute which provided linkage to the `PendingArtifact` related to the download.  This is used by the `ChangeSet`.  I moved this correlation to the `DownloadIterator` and got rid of the `DownloadTask`.